Add API Auth ADR

arottersman opened the pull request:
ADR on how the client app should authenticate its anonymous requests

;tldr/SPOILER ALERT: The client app should have its own user/API token like everyone else. We can't keep it secret, but we can pass it via the client settings in such a way that we can cycle it as frequently as possible. 

Thanks @mmcfarland for coaching me through this, and identifying a lot of the possibilities/trade offs.

@hectcastro — you might want to have a look at this.

Connects #2270

caseycesari commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+To get the token to put in the client's settings, the server can look up the
+token via the client app's `user_id`. The `user_id` will have to remain
+secret so that there's no programmatic way get the token outside of swiping
+it off the client app.
```
I may be missing something, but where is the token stored?

arottersman replied:
The token is stored in the `authtoken_token` table created by [`rest_framework.authtoken`](http://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication)

caseycesari commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+## Consequences
+
+#### Backdoor to the API
+As discussed in Context, allowing the client app to make requests without a user creates a 
```
Minor but consider linking Context to the context section.

caseycesari commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+
+Setting up the app with its own API token will allow us to both
+- cycle the key easily because Django Rest Framework (DRF) handles token
+generation/re-generation, and we already use it for "real" API users
```
How does the cycling happen? Is there an endpoint in DRF to do this?

arottersman replied:
There's a built-in management command `./manage.py drf_create_token -r <username>`. We'll also be exposing an endpoint for use on the new user accounts page.

hectcastro commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+
+##### How the server should get the token
+
+To get the token to put in the client's settings, the server can look up the
```
Is the idea that each user of the application will use a token that is associated to their user ID, or a token specific to the application that we use on their behalf?

At different points while reading through, I thought it was the latter, but then later felt as though the former was being proposed.

arottersman replied:
I'll make this more clear. Trying to propose the latter; we'd be sending a token specific to the application on their behalf. The `user_id` would be that of a user we'd create for the client app to use for all geoprocessing API requests.

When using the app, an individual user's credentials would only be used for the modeling and user endpoints (they're not part of the public API).

arottersman commented:
Just spoke with @mmcfarland offline — going to do a quick experiment to see if we can distinguish the app via session. This is the opposite of what the ADR currently proposes, but may save us a lot of the complexity of setting up users for each app.

hectcastro commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+
+if token is client app:
+
+cache number of requests per IP (like [AnonRateThrottle](http://www.django-rest-framework.org/api-guide/throttling/#anonratethrottle))
```
For this, we just need to make sure the IP it uses is truly the clients vs. the IP of our load balancer, or some other hop internal to our infrastructure. May be hard to test outside of EC2.

hectcastro commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+client app `user_id`'s.
+
+We could:
+1. Use some special email or name like `clientapp09212017` for the app server to look up the `user_id` to get the token. This would add a level
```
I think that this is a good place to start. For example, Raster Foundry has a user that Airflow uses to interact with the main API on its behalf called `rf|airflow-user`.

arottersman commented:
- Per @mmcfarland suggestion, reframed the decision to be more in terms of architectural integrity than security
- After some experiments in detecting what was and wasn't a request from the client app, reworked option 2 to reflect what I found to be possible 
- Added clarification on how the tokens work via DRF
- Added clarification on when the client app would be sending tokens

mmcfarland commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,121 @@
+
+We're creating a [proper, publicly documented API](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/geoprocessing_api/views.py) around our analyze and 
+RWD endpoints: the [Geoprocessing API.](https://github.com/WikiWatershed/model-my-watershed/blob/develop/doc/arch/adr-004-geoprocessing-api.md) The analyze and RWD endpoints have
+have allow-any access; we'd now like to secure them via token authentication
```
"have had", maybe?  Or even, "have always had"?

arottersman replied:
There are a terrifying number of repeated words...I think this was "have always had" before I mangled it.

mmcfarland commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,121 @@
+
+Enforcing token authentication on these previously open endpoints causes
+a problem for our client app; the client app consumes the same API, and
+we want to allow it to use these endpoints un-authenticated. We need a way 
```
Consider rephrasing to illustrate that the endpoints may be authenticated, but unauthenticated users of the app should continue to be able to access them.

Something like "..want to continue to allow unauthenticated users access to these endpoints via the app"

mmcfarland commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,121 @@
+##### When should the client app send its token
+
+For all Geoprocessing API requests, whether there's a logged-in user's token available
+to use or not, ie., we will continue to use a user's credentials only for user/model/project
```
Maybe this is just being pedantic, but technically those requests don't need any credentials to access.  Maybe refer to these as application endpoints, to distinguish them from "api" endpoints?

arottersman commented on doc/arch/adr-005-api-client-app-auth.md:
```diff
@@ -0,0 +1,108 @@
+   - Could be the CSRF token or something else. Might involve some custom
+   generation/management
+   
+1. Some combination of (1) and (2)
```
I removed mention of this, but can add it back in.

mmcfarland replied:
Sorry, I thought I deleted this comment with my others after our discussions.  Still an interesting idea, but didn't intend to muddy the waters.

arottersman commented:
Made the final suggested edits for clarity — going to squash and merge.
